File: Binary_Tree/construct_binary_tree_from_preorder_and_inorder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int preorder_pointer = 0;
    unordered_map<int, int> inorder_position;
    TreeNode* TreeMaker(vector<int>& preorder, vector<int>& inorder, int left, int right){
        if(left > right)
            return NULL;
        if(left == right){
            preorder_pointer++;
            return new TreeNode(inorder[left]);
        }
        int preorder_top_elem = preorder[preorder_pointer];
        int in_pos = inorder_position[preorder_top_elem];
        preorder_pointer++;
        TreeNode* leftNode = TreeMaker(preorder, inorder, left, in_pos - 1);
        TreeNode* rightNode = TreeMaker(preorder, inorder, in_pos + 1, right);
        TreeNode* currentRoot = new TreeNode(preorder_top_elem, leftNode, rightNode); 
        return currentRoot;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int sz = inorder.size();
        for(int i = 0; i < sz; i++)
            inorder_position[inorder[i]] = i;
        return TreeMaker(preorder, inorder, 0, sz - 1);
    }
};

File: Binary_Tree/construct_binary_tree_from_preorder_and_inorder.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# v : preorder pointer

                                                #            v 
                                                # Preorder - 1 2 3 4 5
                                                # Inorder - [3 2 4 (1) 5]    

            #              v                                                            #                    v 
            # Preorder - 1 2 3 4 5                                                      # Preorder - 1 2 3 4 5    
            # Inorder - [3, (2), 4]	                                                    # Inorder - [(5)]            

#                v              #                  v 
# Preorder - 1 2 3 4 5          # Preorder - 1 2 3 4 5
# Inorder - [(3)]               # Inorder - [(4)]         




class Solution(object):
    def __init__(self):
        self.inorder_location = {}
        self.preorder = []
        self.inorder = []
        self.preorder_pointer = 0
    
    def treeBuilder(self, left, right):
        if left > right:
            return None
        
        if left == right:
            newNode = TreeNode(val = self.preorder[self.preorder_pointer])
            self.preorder_pointer += 1
            return newNode
        
        preorder_element = self.preorder[self.preorder_pointer]
        element_inorder_location = self.inorder_location[preorder_element]
        
        self.preorder_pointer += 1
        
        leftSubtree = self.treeBuilder(left, element_inorder_location - 1)
        rightSubtree = self.treeBuilder(element_inorder_location + 1, right)
        currentRoot = TreeNode(val = preorder_element, left=leftSubtree, right=rightSubtree)
        return currentRoot
            
        
    
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        self.preorder = preorder
        self.inorder = inorder
        sz = len(self.preorder)
        for i in range(0, sz):
            self.inorder_location[inorder[i]] = i
        
        return self.treeBuilder(0, sz - 1)
        